cmd/soroban-rpc/lib/preflight/src/preflight.rs

```diff
@@ -0,0 +1,286 @@
+        host_function: invoke_hf_op.host_function,
+        auth: auths.clone(),
+    };
+    let (transaction_data, min_fee) = fees::compute_host_function_transaction_data_and_min_fee(
+        &invoke_host_function_with_auth,
+        &ledger_storage_rc,
+        &storage,
+        &budget,
+        &diagnostic_events,
+        bucket_list_size,
+        ledger_info.sequence_number,
+    )
+    .context("cannot compute resources and fees")?;
+
+    let entries = ledger_storage_rc.get_ledger_keys_requiring_restore();
+    let restore_preamble = if entries.is_empty() {
+        None
+    } else {
+        let read_write_vec: Vec<LedgerKey> = Vec::from_iter(entries);
+        let restore_footprint = LedgerFootprint {
+            read_only: VecM::default(),
+            read_write: read_write_vec.try_into()?,
+        };
+        let (transaction_data, min_fee) =
+            fees::compute_restore_footprint_transaction_data_and_min_fee(
+                restore_footprint,
+                &ledger_storage_rc,
+                bucket_list_size,
+                ledger_info.sequence_number,
+            )
+            .context("cannot compute restore preamble")?;
+        Some(RestorePreamble {
+            transaction_data,
+            min_fee,
+        })
+    };
+
+    Ok(PreflightResult {
+        auth: auths.to_vec(),
+        result: Some(result),
+        transaction_data,
+        min_fee,
+        events: diagnostic_events,
+        cpu_instructions: budget
+            .get_cpu_insns_consumed()
+            .context("cannot get cpu instructions")?,
+        memory_bytes: budget
+            .get_mem_bytes_consumed()
+            .context("cannot get consumed memory")?,
+        restore_preamble,
+    })
+}
+
+fn recorded_auth_payload_to_xdr(payload: &RecordedAuthPayload) -> SorobanAuthorizationEntry {
+    match (payload.address.clone(), payload.nonce) {
+        (Some(address), Some(nonce)) => SorobanAuthorizationEntry {
+            credentials: SorobanCredentials::Address(SorobanAddressCredentials {
+                address,
+                nonce,
+                // signature is left empty. This is where the client will put their signatures when
+                // submitting the transaction.
+                signature_expiration_ledger: 0,
+                signature: ScVal::Void,
+            }),
+            root_invocation: payload.invocation.clone(),
+        },
+        (None, None) => SorobanAuthorizationEntry {
+            credentials: SorobanCredentials::SourceAccount,
+            root_invocation: payload.invocation.clone(),
+        },
+        // the address and the nonce can't be present independently
+        (a,n) =>
+            panic!("recorded_auth_payload_to_xdr: address and nonce present independently (address: {:?}, nonce: {:?})", a, n),
+    }
+}
+
+fn host_events_to_diagnostic_events(events: &Events) -> Vec<DiagnosticEvent> {
+    let mut res: Vec<DiagnosticEvent> = Vec::new();
```

Comment:
Since we know the length of `events.0` you can use [`Vec::with_capacity`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity), which will prevent extra allocations when pushing.

Reply:
Good tip, thanks